### PR TITLE
[G2] Create Cipher light

### LINKED ISSUE #8058: [G2] Create Cipher light
See tentative definition in #8057 - this is similar to the existing MD light (introduced in #7120) and ECP light. (Don't worry about the definition being tentative: it will be refined if needed when we add more users in future tasks.)

This means that the outer guard of `cipher.c` will be changed from `CIPHER_C` to `CIPHER_LIGHT`, and then the parts of the file that are not part of Cipher light will be guarded by `CIPHER_C`. The macro `MBEDTLS_CIPHER_LIGHT` is private and does not appear in `mbedtls_config.h`.

Dependencies in `test_suite_cipher*` will need adjusting.

Rename the existing `all.sh` component `test_full_no_cipher` to `test_full_no_cipher_at_all` and create a new component `test_full_cipher_light_only` that will be initially very similar. For now, enable Cipher light by adding `-DMBEDTLS_CIPHER_LIGHT`.

Add a test `not grep mbedtls_cipher library/cipher.o` to the `no_cipher_at_all` component and a test `not grep mbedtls_cipher_auth library/cipher.o` to the `cipher_light_only` component.

### LINKED ISSUE #8060: [G2] Make CCM and GCM work with the new block_cipher module
- Change `GCM_C` and `CCM_C` from "depends on `CIPHER_C`" to "auto-enables `BLOCK_CIPHER` in `config_adjust_legacy_crypto.h`
- Re-enable those two in `all.sh` component `test_cipher_light_only` (and remove `-DMBEDTLS_BLOCK_CIPHER` from `CFLAGS` it's now auto-enabled).

**Depends on:** #8058

Note: this was prototyped in #8466.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Resolves #8058 following the guide lines defined in #8057 (and also #8337).

Opening as draft PR because the discussion in #8057 is still ongoing.

This might also resolve #8060, but this is just an attempt. Once the PR will be finalized I will update this comment accordingly.

## PR checklist

- [ ] **changelog** post-poned in #8357
- [ ] **backport** not required
- [ ] **tests** provided

### CODE REVIEW

**Comment by valeriosetti** [comment]
For prototype purposes, this PR from this point on will also try to solve #8060

**Comment by valeriosetti** [comment]
InternalCI failed, but the OpenCI is fine, so I would say this PR can be used as prototype for the [ongoing discussion in 8057](https://github.com/Mbed-TLS/mbedtls/pull/8057#discussion_r1379898161)

File: library/cipher_wrap.c [hunk 1]
```diff
@@ -89,12 +89,16 @@ enum mbedtls_cipher_base_index {
...
 #endif
-#if defined(MBEDTLS_CHACHAPOLY_C)
-    MBEDTLS_CIPHER_BASE_INDEX_CHACHAPOLY_BASE,
+#if defined(MBEDTLS_GCM_C) && defined(MBEDTLS_ARIA_C)
+    MBEDTLS_CIPHER_BASE_INDEX_GCM_ARIA,
+#endif
+#if defined(MBEDTLS_GCM_C) && defined(MBEDTLS_CAMELLIA_C)
+    MBEDTLS_CIPHER_BASE_INDEX_GCM_CAMELLIA,
 #endif
+#if defined(MBEDTLS_CIPHER_C)
```
Comment (mpg) [inline library/cipher_wrap.c:101]: I think this could be placed higher - just after Camellia. We don't want cipher light to expose GCM or CCM as part of its API - we just want it to provide the necessary building blocks for `gcm.c` and `ccm.c`.

File: library/cipher_wrap.c [hunk 2]
```diff
@@ -583,22 +585,22 @@ static int gcm_aes_setkey_wrap(void *ctx, const unsigned char *key,
...
 #endif
-#if defined(MBEDTLS_CIPHER_MODE_XTS)
+#if defined(MBEDTLS_CIPHER_MODE_XTS) && defined(MBEDTLS_CIPHER_C)
     NULL,
 #endif
-#if defined(MBEDTLS_CIPHER_MODE_STREAM)
+#if defined(MBEDTLS_CIPHER_MODE_STREAM) && defined(MBEDTLS_CIPHER_C)
     NULL,
 #endif
 #if defined(MBEDTLS_GCM_C)
```
Comment (mpg) [inline library/cipher_wrap.c:606]: I think this should have `&& defined(MBEDTLS_CIPHER_C)` as well.

Comment (mpg) [reply library/cipher_wrap.c:606]: That holds for every instance `GCM_C`, `CCM_C` or `MBEDTLS_CIPHER_HAVE_GCM_AES_VIA_LEGACY_OR_USE_PSA` (whose definition we probably want to change to include the `CIPHER_C` bit, rather than adding it to all occurrences of this macro).

Comment (mpg) [reply library/cipher_wrap.c:606]: (Note: there are instances in `cipher.c` which are concerned as well.)

File: library/cipher_wrap.c (hunk 1 above)
Comment (valeriosetti) [reply library/cipher_wrap.c:101]: Right, when implementing this I kept CCM and GCM there because I knew that AEADs were the functionalities we want to keep with `CIPHER_LIGHT`. However I didn't realized that this kind of support from TLS point of view is achieved by means of "separate" functions like `mbedtls_cipher_aead_encrypt()` which does not go through `cipher_wrap` directly (`cipher_wrap` is still required but it only indirectly through CCM/GCM modules).
Long story short: thanks for pointing this out. I will fix it ;)

File: library/cipher.c [hunk 3]
```diff
@@ -404,9 +290,375 @@ int mbedtls_cipher_setkey(mbedtls_cipher_context_t *ctx,
...
+        if (0 != (ret = mbedtls_cipher_get_base(ctx->cipher_info)->ecb_func(ctx->cipher_ctx,
+                                                                            ctx->operation, input,
+                                                                            output))) {
+            return ret;
+        }
+
+        return 0;
+    }
+
+#if defined(MBEDTLS_GCM_C)
```
Comment (mpg) [inline library/cipher.c:338]: I think this should be guarded with `&& defined(CIPHER_C)` as well.

Comment (valeriosetti) [reply library/cipher.c:338]: I actually skipped that on purpose because I saw `mbedtls_cipher_update` was used by both CCM and GCM so I thought this was necessary for them. Functions like `mbedtls_ccm_crypt` and `mbedtls_ccm_update_ad` use `mbedtls_cipher_update` which eventually returns back to `mbedtls_ccm_update`. I suspect that removing that part might prevent these functions to work properly. Or (option B) we must prevent using cipher functions, so instead of `ccm->cipher->ccm`, we just do `ccm->ccm`.

File: library/cipher_wrap.c (hunk 1 above)
Comment (mpg) [reply library/cipher_wrap.c]: > AEADs were the functionalities we want to keep with `CIPHER_LIGHT`

No they're not :) The functionalities we want to keep with `CIPHER_LIGHT` are those that _are used to build_ GCM and CCM, which is very different. We do _not_ want AEADs as part of the `CIPHER_LIGHT` API, because we don't need them. TLS calls `psa_aead_xxx()` which directly calls `mbedtls_gcm_xxx()` or `mbedtls_ccm_xxx()`, and nothing in the library tries to do GCM or CCM by using `mbedtls_cipher_xxx()` so we want them out of `CIPHER_LIGHT`.

(Said otherwise, we want in `CIPHER_LIGHT` the functions _that are called by_ `mbedtls_gcm_xxx()` / `mbedtls_ccm_xxx()` but we don't want `CIPHER_LIGHT` to call `mbedtls_gcm_xxx()` or `mbedtls_ccm_xxx()`.)

The design document in #8057 says so:
> This excludes:
> - [...]
> - the AEAD/KW API (both one-shot and streaming);

Comment (valeriosetti) [reply library/cipher_wrap.c]: Right. Initially I set guards in cipher module based on what I thought it was required, but I didn't check the TLS code. Once I did yesterday, I immediately realized that you were right ;)

File: library/cipher.c (hunk 3 above)
Comment (mpg) [reply library/cipher.c:338]: And when people call something like `mbedtls_cipher_auth_xxx()` (the AEAD part of the Cipher API), it goes `cipher(GCM)->gcm->cipher(ECB)->aes/aria/camellia`. There's no `gcm->cipher(GCM)` arrow anywhere, and the `cipher(GMC)->gcm` part is not something we need in `CIPHER_LIGHT`.

**Comment by valeriosetti** [comment]
I think that the only test failing in `full_without_ecdhe_ecdsa` on the OpenCI is actually a problem of the CI, because it's passing locally on my PC even using Docker.

**Comment by gilles-peskine-arm** [comment]
That's the second instance of https://github.com/Mbed-TLS/mbedtls/issues/8471 that I've seen. Both in the same build, which may just be a coincidence. Let's keep an eye on this, please report any other instances of intermittent test failures that you notice.

**Comment by mpg** [comment]
Superseded by #8519.
